### PR TITLE
add CUDA extension

### PR DESCRIPTION
**Pull request by CarloLucibello**
Building on https://github.com/FluxML/NNlib.jl/pull/492, this removes CUDA as a dependency and implements option (1) in #2265, meaning that the CUDA functionality is now unlocked with the explicit double import `using CUDA, cuDNN` in users' code.

Other notes:
- Minimum supported julia version becomes 1.9
- This is a breaking change
- Moving to option (2) in #2265  requires minimal changes to this PR and the creation of a `FluxCUDA` package.
- I didn't add/remove/change any test, just moved them around
- I'm closely following the structure of the AMDGPU extension

### CODE REVIEW

**Comment by ToucheSir** [comment]
I'm pretty ambivalent on whether 1) or 2) would be better, but 2) might let us keep compat with older Julia versions if we made it load NNlibCUDA and/or the Flux CUDA code on <1.9.

**Comment by CarloLucibello** [comment]
> I'm pretty ambivalent on whether 1) or 2) would be better, but 2) might let us keep compat with older Julia versions if we made it load NNlibCUDA and/or the Flux CUDA code on <1.9.

let's discuss this in #2265

**Comment by CarloLucibello** [comment]
@ToucheSir your suggestion works! This should be ready to merge

File: NEWS.md [hunk 1]
```diff
@@ -2,6 +2,12 @@
 
 See also [github's page](https://github.com/FluxML/Flux.jl/releases) for a complete list of PRs merged before each release.
 
+## v0.14.0
+* Flux now requires julia v1.9 or later.
+* CUDA.jl is not a hard dependency anymore. CUDA support is now provided through the extension mechanism. In order to unlock the CUDA 
+functionalities user are required to load CUDA, e.g. with `using CUDA`.
+The package `cuDNN.jl` also needs to be installed in the environment.
```
Comment (CarloLucibello) [inline NEWS.md:9]: @ToucheSir is this phrasing ok?

File: ext/FluxCUDAExt/FluxCUDAExt.jl [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    Flux.CUDA_LOADED[] = true
+
+    try
+       Base.require(Flux, :cuDNN)
+    catch
+        @warn """Package cuDNN not found in current path.
+        - Run `import Pkg; Pkg.add(\"cuDNN\")` to install the cuDNN package, then restart julia.
+        - If cuDNN is not installed, some Flux functionalities will not be available when running on the GPU.
+        """
+    end
```
Comment (CarloLucibello) [inline ext/FluxCUDAExt/FluxCUDAExt.jl:46]: We could also be more strict and `@error` instead, but I think a warning is ok.

**Comment by CarloLucibello** [comment]
This should be ready to go. After merging this I will wait a couple of days, do some local testing, then tag the v0.14 release. Cc @ToucheSir @mcabbott @darsnack

File: NEWS.md (hunk 1 above)
Comment (ToucheSir) [reply NEWS.md:9]: Yes, just some wordsmithing for your consideration:
```suggestion
* Flux now requires julia v1.9 or later.
* CUDA.jl is no longer a hard dependency. CUDA support is now provided through the [extension](https://pkgdocs.julialang.org/v1/creating-packages/#Conditional-loading-of-code-in-packages-(Extensions)) mechanism.
In order to unlock CUDA functionality, the `CUDA` and `cuDNN` packages must be installed in the environment and CUDA.jl must be loaded, e.g. with `using CUDA`.
```

File: ext/FluxCUDAExt/FluxCUDAExt.jl [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+include("functor.jl")
+
+function __init__()
+    Flux.CUDA_LOADED[] = true
+
+    try
+       Base.require(Flux, :cuDNN)
+    catch
+        @warn """Package cuDNN not found in current path.
```
[the remark this answers is not in the document]
Comment (ToucheSir) [reply ext/FluxCUDAExt/FluxCUDAExt.jl:42]: Ok, tried some ideas locally and this appears to work:
```julia
try
    Base.require(Main, :cuDNN)
```
Not sure how general this is, but I haven't encountered any issues yet.

[the remark this answers is not in the document]
Comment (CarloLucibello) [reply ext/FluxCUDAExt/FluxCUDAExt.jl:42]: This works!

**Comment by CarloLucibello** [comment]
@ToucheSir everything seems fine now. Should we pull the trigger?

**Review by darsnack — APPROVED** [review]

**Comment by darsnack** [comment]
No complaints from me

**Review by ToucheSir — APPROVED** [review]

**Comment by IanButterworth** [comment]
Awesome!

**Pull request merged**